### ISSUE TITLE
[Feature] Support `yarnclean` functionality via resolutions or config

### ISSUE THREAD

**Issue by darthtrevino** [yarnpkg/berry#1629]
- [x] I'd be willing to implement this feature (with some guidance perhaps, I'm not familiar with the berry codebase)
- [ ] This feature can already be implemented through a plugin

**Describe the user story**

Some important dependencies have malformed dependency lists and bring in transitive packages that can break builds. The example I have in mind is `@types/styled-components`, which brings in `@types/react-native`, which clashes with DOM typings built into TypeScript.

In Yarn v1, I could add `@types/react-native` to `.yarnclean`, and that dependency would not be be installed. This only makes a _single workspace_ work though. 

**Describe the solution you'd like**

Perhaps there's a way to block dependencies from being included in the _current package and downstream dependencies_

e.g.

```json
"dependencies": {
   "@types/styled-components": "^5.3.1",
},
"resolutions": {
   "@types/react-native": null
}
```

or 

```json
"dependencies": {
   "@types/styled-components": "^5.3.1",
},
"excludeDependencies": ["@types/react-native"]
]
```

**Describe the drawbacks of your solution**

An extension like this would be yarn-specific, and npm clients would still have that extension drawn into their workspace.

**Additional context**

https://github.com/DefinitelyTyped/DefinitelyTyped/issues/33015

**Comment by merceyz** [comment]
I guess you could try with aliasing it to the normal react types.
```
"resolutions": {
   "@types/react-native": "@types/react@*"
}
```
Is this a problem in both PnP and node_modules or?

**Comment by darthtrevino** [comment]
My use case is mostly with node_modules, but this affects both linkers

**Comment by darthtrevino** [comment]
I fat-fingered that edit - the workaround did not work

**Comment by merceyz** [comment]
I didn't write it correctly, updated, try again please

**Comment by darthtrevino** [comment]
Tried the update, no effect

**Comment by merceyz** [comment]
Do you have a reproduction for that? Setting that resolution and running `yarn add @types/react-native` shows that `node_modules/@types/react-native` contains the source for `@types/react`

**Comment by darthtrevino** [comment]
https://github.com/darthtrevino/yarn-berry-1629

I do see that if I move the "resolutions" field to the top-level monorepo root, then that does what you're suggesting. I'll try that out in my app

**Comment by merceyz** [comment]
> I do see that if I move the "resolutions" field to the top-level monorepo root, then that does what you're suggesting

Yes, resolutions only work from the the top level `package.json`, Yarn also tells you this when you run an install
```
$ yarn
➤ YN0000: ┌ Validation step
➤ YN0057: │ a: Resolutions field will be ignored
➤ YN0000: └ Completed
➤ YN0000: ┌ Resolution step
➤ YN0000: └ Completed
➤ YN0000: ┌ Fetch step
➤ YN0000: └ Completed
➤ YN0000: ┌ Link step
➤ YN0000: └ Completed
➤ YN0000: Done with warnings in 0.16s
```

**Comment by darthtrevino** [comment]
lol, derp - I must've missed that

**Comment by darthtrevino** [comment]
The workaround did work in my projects - it's not the most straightforward of configs, but it lets me use the types I want to use. I'll leave this issue open as a suggestion for more direct approaches to problems like that
